architecture/2025-02-user-defined-resource-type-feature-spec.md

```diff
@@ -0,0 +1,1212 @@
+```
+terraform {
+  required_providers {
+    radius = {
+      source  = "terraform.radapp.io"
+      version = "~> 0.46"
+    }
+  }
+}
+
+variable "context" {
+  description = "This variable contains Radius recipe context."
+  type = any
+}
+
+resource "radius_applications_core_containers" "proxy" {
+  name                = "proxy"
+  environment         = var.context.environment.id
+  application         = var.context.application
+  container_image     = "envoy:latest"
+}
+
+resource "radius_applications_core_containers" "serviceContainer" {
+  name                = "${var.context.resource.name}-container"
+  environment         = var.context.environment.id
+  application         = var.context.application.id
+  container_image     = var.context.resource.properties.container.image
+  ...
+}
+```
+
+**User Experience 2 – Using Bicep template**
+
+The recipe for the service resource type will leverage Radius resource types.
+
+**`webservice-recipe.bicep`**:
+
+```yaml
+extension radiusCore
+
+@description('Information about what resource is calling this Recipe. Generated by Radius.')
+param context object
+
+resource proxy 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'proxy'
+  properties: [
+    environment: context.environment.id
+    application: context.application.id
+    container: {
+      image: 'envoy:latest'
+      ...
+    },
+  ]
+}
+
+resource serviceContainer 'Applications.Core/containers@2023-10-01-preview' = {
+  name: '${context.resource.name}-container'
+  properties: {
+    application: context.application.id
+    container: context.resource.properties.container
+    ...
+  }
+}
+```
+
+### **User Story 6 – Conditional resources** 
+
+As a platform engineer, I need to write a recipe which creates additional resources dependent upon the properties set by the developer.
+
+**Summary**
+
+The platform engineer adds a boolean property to the `webservice` resource type from the previous user story.
+
+**`service-resource-type.bicep`**:
+
+```diff
+namespace: Radius.Resources
+types:
+  webService:
+    apiVersions:
+      '2023-10-01-preview':
+        schema: 
+          type: object
+          properties:
+            ...
++            ingress:
++              type: boolean
++              description: "Optional: When true, provisions an L7 ingress gateway mapped to the port named http"
+            container:
+              type: object
+              properties:
+                image:
+                  type: string
+                  description: "Required: The container image"
+                ...
+```
+
+**User Experience 1 – Using Terraform configuration**
+
+```diff
+terraform {
+  required_providers {
+    radius = {
+      source  = "terraform.radapp.io"
+      version = "~> 0.46"
+    }
+  }
+}
+
+variable "context" {
+  description = "This variable contains Radius recipe context."
+  type = any
+}
+
++ if var.context.resource.properties.ingress {
++  resource "applications_core_gateways" "gateway" {
```

Comment:
How would we create UDT resources using Terraform Radius Provider? Each time a new resource type is registered, we would need to update the provider somehow.